torch/functional.py

```diff
@@ -1674,6 +1675,100 @@ def norm(input, p: Optional[Union[float, str]] = "fro", dim=None, keepdim=False,
+
+        shape (int, sequence of ints, or torch.Size): The shape of the arbitrary
+            tensor.
+
+    Returns:
+        tuple of Tensors: Each ``i``-th tensor in the ouput corresponds with
+        dimension ``i`` of :attr:`shape`. Each tensor has the same shape as
+        ``indices`` and contains one index into dimension ``i`` for each of the
+        flat indices given by ``indices``.
+
+    Example::
+
+        >>> import torch
+        >>> torch.unravel_index(torch.tensor(4), (3, 2))
+        (tensor(2),
+         tensor(0))
+
+        >>> torch.unravel_index(torch.tensor([4, 1]), (3, 2))
+        (tensor([2, 0]),
+         tensor([0, 1]))
+
+        >>> torch.unravel_index(torch.tensor([0, 1, 2, 3, 4, 5]), (3, 2))
+        (tensor([0, 0, 1, 1, 2, 2]),
+         tensor([0, 1, 0, 1, 0, 1]))
+
+        >>> torch.unravel_index(torch.tensor([1234, 5678]), (10, 10, 10, 10))
+        (tensor([1, 5]),
+         tensor([2, 6]),
+         tensor([3, 7]),
+         tensor([4, 8]))
+
+        >>> torch.unravel_index(torch.tensor([[1234], [5678]]), (10, 10, 10, 10))
+        (tensor([[1], [5]]),
+         tensor([[2], [6]]),
+         tensor([[3], [7]]),
+         tensor([[4], [8]]))
+
+        >>> torch.unravel_index(torch.tensor([[1234], [5678]]), (100, 100))
+        (tensor([[12], [56]]),
+         tensor([[34], [78]]))
+    """
+    res_tensor = _unravel_index(indices, shape)
+    return res_tensor.unbind(-1)
+
+def _unravel_index(indices: Tensor, shape: Union[int, Sequence[int], torch.Size]) -> Tensor:
+    torch._check_type(
+        not indices.is_complex() and not indices.is_floating_point() and not indices.dtype == torch.bool,
+        lambda: f"expected 'indices' to be integer dtype, but got {indices.dtype}")
+
+    torch._check_type(
+        isinstance(shape, (int, Sequence, torch.Size)),
+        lambda: f"expected 'shape' to be int, sequence of ints, or torch.Size, but got {type(shape)}")
+
+    if isinstance(shape, int):
+        shape_ = torch.Size([shape])
+    elif isinstance(shape, Sequence):
+        for dim in shape:
+            torch._check_type(
+                isinstance(dim, int),
+                lambda: f"expected 'shape' sequence to only contain ints, but got {type(dim)}")
+        shape_ = torch.Size(shape)
+    else:
+        shape_ = shape
+
+    shape_tensor = torch.tensor(shape_ + torch.Size([1]), device=indices.device, dtype=torch.int64)
+
+    torch._check_tensor_all(
+        shape_tensor >= 0,
+        lambda: "'shape' cannot have negative values")
+
+    shape_numel = shape_tensor.prod()
+
+    # This conversion to int64 is to support the case where `shape_numel` is
+    # greater than the largest value of `indices.dtype`. The comparison
+    # `indices < shape_numel` fails in that case.
+    indices_ = indices.to(torch.int64)
+
+    torch._check_tensor_all(
+        (indices_ >= 0).all() & (indices_ < shape_numel).all(),
+        lambda: f"invalid index, must be between 0 and {shape_numel.item() - 1}")
```

Comment:
This synchronizes, so we might want to avoid them. These are the reason why the FakeTensor test fails, and this would disallow this function to be traced through in torch.compile. Perhaps we just want to document that the indices should all be of the right size and move on.

Reply:
Done. I also removed the check for the elements of `shape` and mentioned in the documentation that they should all be non-negative